tests/test_dtypes.py

```diff
@@ -0,0 +1,328 @@
+        # Single precision have an 'f' suffix (half is promoted to single)
+        call_str = '%sf' % call_str
+
+    # Operator setup
+    a = Symbol(name='a', dtype=dtype)
+    b = Scalar(name='b', dtype=dtype)
+
+    eq = Eq(a, sym(b))
+    op = Operator(eq)
+
+    # Ensure the generated function call has the correct form
+    assert call_str in str(op)
+
+
+@pytest.mark.parametrize('dtype', [np.complex64, np.complex128])
+def test_complex_override(dtype: np.dtype[np.complexfloating]) -> None:
+    """
+    Tests overriding complex values in op.apply().
+    """
+
+    grid = Grid(shape=(5, 5))
+    x, y = grid.dimensions
+
+    c = Constant(name='c', dtype=dtype, value=1.0 + 0.0j)
+    u = Function(name='u', grid=grid, dtype=dtype)
+    eq = Eq(u, x * x.spacing + c * y * y.spacing)
+    op = Operator(eq)
+    op.apply(c=2.0 + 1.0j)
+
+    # Check against numpy result
+    dx, dy = grid.spacing_map.values()
+    xx, yy = np.meshgrid(np.linspace(0, 4, 5, dtype=dtype),
+                         np.linspace(0, 4, 5, dtype=dtype))
+    expected = xx * dx + yy * dy * dtype(2.0 + 1.0j)
+    assert np.allclose(u.data.T, expected)
+
+
+def test_half_time_deriv() -> None:
+    """
+    Tests taking the time derivative of a float16 function.
+    """
+
+    grid = Grid(shape=(5, 5))
+    x, y = grid.dimensions
+    t = grid.time_dim
+
+    f = TimeFunction(name='f', grid=grid, space_order=2, dtype=np.float16)
+    g = Function(name='g', grid=grid, dtype=np.float16)
+    eqns = [Eq(f.forward, t * x * x.spacing +
+               y * y.spacing),
+            Eq(g, f.dt)]
+    op = Operator(eqns)
+    op.apply(time=10, dt=1.0)
+
+    # Check against expected result
+    dx = grid.spacing_map[x.spacing]
+    xx = np.repeat(np.linspace(0, 4, 5, dtype=np.float16)[np.newaxis, :], 5, axis=0)
+    expected = xx * np.float16(dx)
+    assert np.allclose(g.data.T, expected)
+
+
+@pytest.mark.parametrize('dtype', [np.complex64, np.complex128])
+def test_complex_time_deriv(dtype: np.dtype[np.complexfloating]) -> None:
+    """
+    Tests taking the time derivative of a complex-valued function.
+    """
+
+    grid = Grid(shape=(5, 5))
+    x, y = grid.dimensions
+    t = grid.time_dim
+
+    f = TimeFunction(name='f', grid=grid, space_order=2, dtype=dtype)
+    g = Function(name='g', grid=grid, dtype=dtype)
+    eqns = [Eq(f.forward, t * x * x.spacing * (1.0 + 0.0j) +
+               t * y * y.spacing * (0.0 + 1.0j)),
+            Eq(g, f.dt)]
+    op = Operator(eqns)
+    op.apply(time=10, dt=1.0)
+
+    # Check against expected result
+    dx, dy = grid.spacing_map.values()
+    xx, yy = np.meshgrid(np.linspace(0, 4, 5, dtype=dtype),
+                         np.linspace(0, 4, 5, dtype=dtype))
+    expected = xx * dx + yy * dy * dtype(0.0 + 1.0j)
+    assert np.allclose(g.data.T, expected)
+
+
+@pytest.mark.parametrize('dtype', [np.complex64, np.complex128])
+def test_complex_space_deriv(dtype: np.dtype[np.complexfloating]) -> None:
```

Comment:
Test derivatives on grids with complex spacing?

Reply:
Hm? Is this a suggestion for rewriting the docstring? Or you mean that should be a test of its own? That's essentially what this case covers

Reply:
Oh yeah, must have misread it. Ignore